### PR TITLE
8316660: (fs) Files.lines implementation can use MemorySegment for file mapping

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes using a `MemorySegment` instead of a `ByteBuffer` in the `Files::lines` implementation (in the internal class `FileChannelLinesSpliterator`).

The old solution deterministically unmapped the `ByteBuffer` if closed. If not closed, the `ByteBuffer` would be GC:ed.

This proposal mimics the old behavior but may hold the mapped memory region slightly longer.

Tested and passed tier1, tier2, tier3

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316660](https://bugs.openjdk.org/browse/JDK-8316660): (fs) Files.lines implementation can use MemorySegment for file mapping (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15814/head:pull/15814` \
`$ git checkout pull/15814`

Update a local copy of the PR: \
`$ git checkout pull/15814` \
`$ git pull https://git.openjdk.org/jdk.git pull/15814/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15814`

View PR using the GUI difftool: \
`$ git pr show -t 15814`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15814.diff">https://git.openjdk.org/jdk/pull/15814.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15814#issuecomment-1729328029)

### CODE REVIEW

File: src/java.base/share/classes/java/nio/file/FileChannelLinesSpliterator.java [hunk 1]
```diff
@@ -190,9 +196,9 @@ private String readLine() {
         }
     }
 
-    private ByteBuffer getMappedByteBuffer() {
+    private MemorySegment getMappedByteBuffer() {
```
Comment (liach) [inline src/java.base/share/classes/java/nio/file/FileChannelLinesSpliterator.java:199]: This method's name should be updated now that the return type has changed.

File: test/jdk/java/nio/file/FileChannelLinesSpliterator/Parallel.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.LockSupport;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class Parallel {
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/FileChannelLinesSpliterator/Parallel.java:45]: This test should needs to move to test/jdk/java/nio/file/Files and be named something like LinesParallel. The description can also say that it's a test for Files.lines.

Comment (AlanBateman) [reply test/jdk/java/nio/file/FileChannelLinesSpliterator/Parallel.java:45]: I also changed the JBS issue to make it clear that this is about the Files.lines implementation, do you mind adjusting the PR to align with that as "FileChannelLinesSplitator" is an internal class.

File: src/java.base/share/classes/java/nio/file/Files.java [hunk 3]
```diff
@@ -4135,10 +4135,10 @@ private static Stream<String> createFileChannelLinesStream(FileChannel fc, Chars
             // for a non-zero length file so disallow this case.
             if (length > 0 && length <= Integer.MAX_VALUE) {
                 FileChannelLinesSpliterator fcls =
-                    new FileChannelLinesSpliterator(fc, cs, 0, (int) length);
+                    new FileChannelLinesSpliterator(fc, cs,(int) length);
```
Comment (bplb) [inline src/java.base/share/classes/java/nio/file/Files.java:4138]: Suggestion: `, (int) length` -> `, (int)length`.

File: src/java.base/share/classes/java/nio/file/FileChannelLinesSpliterator.java [hunk 4]
```diff
@@ -70,46 +72,33 @@
 final class FileChannelLinesSpliterator implements Spliterator<String> {
 
     static final Set<Charset> SUPPORTED_CHARSETS = Set.of(
-        UTF_8.INSTANCE,
-        ISO_8859_1.INSTANCE,
-        US_ASCII.INSTANCE
+            UTF_8.INSTANCE,
```
Comment (bplb) [inline src/java.base/share/classes/java/nio/file/FileChannelLinesSpliterator.java:75]: Suggestion: change indentation of these three set elements back to what it was.

**Comment by AlanBateman** [comment]
There is significant refactoring here that will take time to review, one thing that will need to be examined is the use of a shared Arena as I think the close op becomes O(N) on the number of platform threads.

**Comment by minborg** [comment]
Keep alive.

**Pull request closed without merging**